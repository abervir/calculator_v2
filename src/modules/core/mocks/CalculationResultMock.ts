import { OperationIcons } from '@modules/calculator/enums';
import { CalculationResult } from '@modules/calculator/models';

export const CalculationResultMock: CalculationResult = {
  id: 1,
  result: 1,
  operands: [0, 1],
  operation: {
    id: 'ID',
    title: 'Operation1',
    icon: OperationIcons.PLUS,
    sign: '+',
    operation: (a: number, b: number) => a + b,
  },
};
