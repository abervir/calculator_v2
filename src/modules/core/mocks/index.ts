export * from './CalculationResultMock';
export * from './EventMock';
