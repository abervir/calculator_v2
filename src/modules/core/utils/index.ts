export * from './input.utils';
export * from './shared.utils';
export * from './math.utils';
