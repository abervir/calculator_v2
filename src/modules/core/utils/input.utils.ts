import { KeyboardEvent } from 'react';
import { ALLOWED_INPUT_CHARS } from '../constants';

export class InputUtils {
  static isInputNaN(event: KeyboardEvent<HTMLInputElement>): boolean {
    return (
      isNaN(Number(event.key)) &&
      !ALLOWED_INPUT_CHARS.includes(event.key) &&
      !event.ctrlKey
    );
  }

  static onKeyPressChecker(event: KeyboardEvent<HTMLInputElement>): void {
    if (InputUtils.isInputNaN(event)) {
      event.preventDefault();
    }
  }
}
