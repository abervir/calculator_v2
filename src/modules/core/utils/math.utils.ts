export class MathUtils {
  static getHighestPrimeNumberBetweenTwoNumbers(
    minValue: number,
    maxValue: number
  ): number | null {
    let primeNumber = null;

    for (var i = minValue; i <= maxValue; i++) {
      if (MathUtils.isPrime(i)) {
        primeNumber = i;
      }
    }

    return primeNumber;
  }

  static isPrime(value: number): boolean {
    if (value < 2) {
      return false;
    }

    for (let i = 2; i < value; i++) {
      if (value % i === 0) {
        return false;
      }
    }

    return true;
  }

  static remainderOfDivision(a: number, b: number): number {
    return a % b;
  }

  static divide(a: number, b: number): number {
    return a / b;
  }

  static sum(a: number, b: number): number {
    return a + b;
  }
}
