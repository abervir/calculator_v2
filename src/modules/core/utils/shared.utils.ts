export class SharedUtils {
  static copyTextToClipboard(value: string): void {
    navigator.clipboard.writeText(value);
  }
}
