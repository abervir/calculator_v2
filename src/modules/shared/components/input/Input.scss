@import '../../../../assets/styles/variables.scss';

.calc-input {
  border-bottom: $border;

  input {
    border: none;
    width: 100%;
    outline: none;
    padding: 5px;
    box-sizing: border-box;
  }
}
