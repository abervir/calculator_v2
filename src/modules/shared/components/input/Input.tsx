import React, { KeyboardEvent } from 'react';
import { InputUtils } from '@modules/core/utils';

import './Input.scss';

interface Props {
  placeholder?: string;
  isNumbersOnly?: boolean;
  onChange: (value: string) => void;
}

export const Input = (props: Props) => {
  const { placeholder = 'Value', isNumbersOnly, onChange } = props;

  const onKeyPressChecker = (event: KeyboardEvent<HTMLInputElement>): void => {
    if (!isNumbersOnly) {
      return;
    }

    InputUtils.onKeyPressChecker(event);
  };

  return (
    <div className='calc-input'>
      <input
        placeholder={placeholder}
        onChange={({ target }) => onChange(target.value)}
        onKeyDown={onKeyPressChecker}
      />
    </div>
  );
};
