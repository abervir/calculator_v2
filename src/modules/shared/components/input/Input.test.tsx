/* eslint-disable jest/no-conditional-expect */
import React from 'react';
import { shallow } from 'enzyme';

import { EventMock } from '@modules/core/mocks';
import { InputUtils } from '@modules/core/utils';

import { Input } from './Input';

const placeholder = 'placeholder';
const isNumbersOnly = false;
const onChange = jest.fn((_: string) => {});

describe('Input', () => {
  it('should render component', () => {
    expect(
      shallow(
        <Input
          placeholder={placeholder}
          isNumbersOnly={isNumbersOnly}
          onChange={onChange}
        />
      )
    ).toMatchSnapshot();
  });

  it('should invoke onChange function after typing input', () => {
    const wrapper = shallow(
      <Input
        placeholder={placeholder}
        isNumbersOnly={isNumbersOnly}
        onChange={onChange}
      />
    );
    wrapper.find('input').simulate('change', EventMock);

    expect(onChange).toHaveBeenCalled();
  });

  [
    {
      isNumbersOnly: false,
    },
    {
      isNumbersOnly: true,
    },
  ].forEach(({ isNumbersOnly }) => {
    const shouldInvoke = isNumbersOnly ? '' : 'not';
    it(`should ${shouldInvoke} invoke onKeyPressCheckerSpy on KeyDown if isNumbersOnly is set to ${isNumbersOnly}`, () => {
      const wrapper = shallow(
        <Input
          placeholder={placeholder}
          isNumbersOnly={isNumbersOnly}
          onChange={onChange}
        />
      );
      const onKeyPressCheckerSpy = jest
        .spyOn(InputUtils, 'onKeyPressChecker')
        .mockImplementation();
      wrapper.find('input').simulate('keydown', new KeyboardEvent('keydown'));

      isNumbersOnly
        ? expect(onKeyPressCheckerSpy).toHaveBeenCalledWith(
            new KeyboardEvent('keydown')
          )
        : expect(onKeyPressCheckerSpy).not.toHaveBeenCalled();
    });
  });
});
