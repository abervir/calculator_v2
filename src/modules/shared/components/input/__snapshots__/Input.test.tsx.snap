// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render component 1`] = `
<div
  className="calc-input"
>
  <input
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="placeholder"
  />
</div>
`;
